File: web/frontend/src/textOrder.js
// HOW MUCH OF THE ORIGINAL A CARD SHOWS, and who decides.
//
// THE OWNER'S SPEC: "There will be per work control over whether the cards are to
// show 1) translations above quotations, 2) quotations above translation, 3) no
// translation, 4) no quotations. same control will be there in metadata section on
// per language basis. the work controls will supercede the metadata controls."
//
// IT REPLACES A GUESS WITH AN INSTRUCTION. What shipped this morning inferred the
// answer from a `readLanguages` list — name the languages you read, and a quote in
// any other one led with its translation. That was one bit where the reader wanted
// four, and it could not say "never show me the Bengali" or "never show me the
// English" at all. The owner's own ruling on being shown the overlap: the four
// states absorb the declaration, because "I read German" IS "for German, the
// quotation goes above the translation" and the other two states have no spelling
// in a boolean.
//
// THE FOUR ARE ONE AXIS, WHICH IS WHY A SLIDER IS THE RIGHT CONTROL — the owner
// asked for one ("I can slide across the 4 options beside it"), and the reason it
// works is that they are ordered: each stop shows strictly more of the original
// than the last. That is the order they are declared in, and the order the row
// draws them in.
export const TEXT_ORDERS = ['trans-only', 'trans-first', 'quote-first', 'quote-only']

// WHAT EACH ONE MEANS, in the owner's words, so the table's labels and this
// module cannot drift apart:
//
//   trans-only   "no quotations"              — the translation, alone
//   trans-first  "translations above quotations"
//   quote-first  "quotations above translation"
//   quote-only   "no translation"             — the quotation, alone
//
// AND `quote-first` IS THE DEFAULT, because it is what the app did before any of
// this existed: every quote reads as written, with its translation underneath. A
// default that reordered a library on upgrade would be the app changing somebody's
// cards for a preference they never expressed.
export const TEXT_ORDER_DEFAULT = 'quote-first'

// INHERIT IS THE ABSENCE OF A VALUE, not a fifth stop. A work with no opinion
// stores nothing, so the language decides; a language with no opinion stores
// nothing, so the master does. Storing an explicit "inherit" would mean two
// spellings of the same state and a migration the first time one of them changed.
const known = (v) => (TEXT_ORDERS.includes(v) ? v : null)

// resolveTextOrder — the precedence, in one place, top wins:
//
//   1. the WORK (a book, a film, a show, a game) or, for a standalone quote, its
//      BOARD. The owner's ruling: a quote has no work row — `work_title` is a
//      plain string on the quote, so two quotes can spell one work differently —
//      and a board is the nearest thing it has to a container.
//   2. the LANGUAGE the line is in.
//   3. the MASTER, which is the slider above the column.
//
// `scope` is whichever of work-or-board applies; the caller knows which it has and
// this does not need to. One argument rather than two, because a card belongs to
// exactly one of them and a function taking both would have to decide which wins —
// a decision neither the owner nor the schema has any use for.
export function resolveTextOrder({ scope, language, byLanguage, master } = {}) {
  const own = known(scope)
  if (own) return own
  const lang = foldLanguage(language)
  const byLang = lang && byLanguage ? known(byLanguage[lang]) : null
  if (byLang) return byLang
  return known(master) || TEXT_ORDER_DEFAULT
}

// The same fold the rest of the app matches language names with — a language is a
// name the reader typed, so "ENG", "eng" and " Eng " are one language.
export function foldLanguage(v) {
  return String(v || '').trim().toLowerCase()
}

// textOrderFromReadLanguages — the retirement of `readLanguages`, and it has to be
// EXACTLY today's behaviour or it is a change nobody asked for.
//
// What that preference meant: an empty list read everything, so every quote read as
// written; a non-empty list read those languages as written and led with the
// translation for every other. So:
//
//   empty  -> master quote-first, nothing per language. Every card as before.
//   listed -> master trans-first (what an undeclared language did), and each
//             declared language quote-first (what a declared one did).
//
// A LANGUAGE WITH NO NAME CANNOT BE DECLARED, so a blank entry is dropped rather
// than folded to "" and given a row that would then out-rank the master for every
// untagged quote in the library.
export function textOrderFromReadLanguages(readLanguages) {
  let list = []
  try {
    const parsed = typeof readLanguages === 'string' ? JSON.parse(readLanguages) : readLanguages
    if (Array.isArray(parsed)) list = parsed
  } catch {
    list = []
  }
  const names = list.map(foldLanguage).filter(Boolean)
  if (!names.length) return { master: 'quote-first', byLanguage: {} }
  const byLanguage = {}
  for (const n of names) byLanguage[n] = 'quote-first'
  return { master: 'trans-first', byLanguage }
}

// masterIsCustom — whether the slider above the column should draw dimmed.
//
// THE OWNER'S: "when other knobs are adjusted (custom), it will lose contrast,
// which will indicate custom state." So the master is three things at once — a
// stored default a new language inherits, a bulk setter that pushes every row to
// itself, and an indicator of whether the rows still agree with it. This answers
// the third. A row that agrees with the master is not custom, however it got there.
export function masterIsCustom(master, byLanguage) {
  const m = known(master) || TEXT_ORDER_DEFAULT
  return Object.values(byLanguage || {}).some((v) => known(v) && v !== m)
}

File: internal/httpapi/text_order.go
package httpapi

import (
	"encoding/json"
	"sort"
	"strings"
)

// HOW MUCH OF THE ORIGINAL A CARD SHOWS, stored.
//
// THE OWNER'S FOUR STATES: "1) translations above quotations, 2) quotations above
// translation, 3) no translation, 4) no quotations" — set per language here, and
// per work or per board on the row itself, where the work's answer wins.
//
// IT REPLACES `readLanguages`, on the owner's ruling. That preference held a list
// of languages the reader can read and inferred the rest: declared meant "as
// written", undeclared meant "translation first". Four states say both of those
// and two more the list could not spell at all, so the list is absorbed rather
// than kept beside them. It is still READ, once, to migrate an account that has
// one — the client does that, because the client is what knows whether this
// account has ever opened the new table.
//
// THE ORDER OF THE FOUR IS ONE AXIS and matches web/frontend/src/textOrder.js
// exactly: each shows strictly more of the original than the last. That is what
// makes the reader's control a slider rather than four buttons, and a server that
// disagreed about the order would still validate every value while the client drew
// them in a different sequence.
var textOrders = []string{"trans-only", "trans-first", "quote-first", "quote-only"}

// textOrderDefault is what the app did before any of this: every quote reads as
// written, with its translation underneath. Stored blobs omit it rather than spell
// it, so a reader who has changed nothing has no row and no master.
const textOrderDefault = "quote-first"

// textOrderMax bounds the per-language map, for the reason readLanguagesMax bounds
// its list: a preference is not a corpus.
const textOrderMax = 64

func validTextOrder(s string) bool {
	for _, o := range textOrders {
		if o == s {
			return true
		}
	}
	return false
}

// textOrderPrefs is the shape the blob holds. `Master` is the slider above the
// column; `ByLanguage` is the rows under it, keyed by folded language name.
type textOrderPrefs struct {
	Master     string            `json:"master,omitempty"`
	ByLanguage map[string]string `json:"byLanguage,omitempty"`
}

// normalizeTextOrder is the round trip loadPrefs and the PUT handler both use.
//
// IT DROPS WHAT MEANS NOTHING RATHER THAN REFUSING THE BLOB, with one exception:
// a value that is not one of the four is a client sending a state this server has
// never heard of, and accepting it would store something no reader can see or
// clear. So an unknown STATE is a 400, while a master equal to the default and a
// row equal to the master are simply dropped — they are true without being stored,
// and keeping them would make "has this reader changed anything" unanswerable.
//
// AND THE KEYS ARE SORTED, so the bytes are stable: setting a language and unsetting
// it again produces the blob it started with, and a diff of two accounts means
// something. Same rule normalizeReadLanguages and cleanDirections follow.
func normalizeTextOrder(raw string) (string, bool) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", true
	}
	var in textOrderPrefs
	if err := json.Unmarshal([]byte(raw), &in); err != nil {
		return "", false
	}
	if len(in.ByLanguage) > textOrderMax {
		return "", false
	}
	out := textOrderPrefs{ByLanguage: map[string]string{}}
	master := textOrderDefault
	if in.Master != "" {
		if !validTextOrder(in.Master) {
			return "", false
		}
		master = in.Master
		if master != textOrderDefault {
			out.Master = master
		}
	}
	keys := make([]string, 0, len(in.ByLanguage))
	for name := range in.ByLanguage {
		keys = append(keys, name)
	}
	sort.Strings(keys)
	for _, name := range keys {
		key := foldLanguageName(name)
		if key == "" || len([]rune(key)) > languageNameMaxRunes {
			return "", false
		}
		state := in.ByLanguage[name]
		if !validTextOrder(state) {
			return "", false
		}
		// A ROW THAT AGREES WITH THE MASTER IS NOT A SETTING. The reader's slider
		// above the column pushes every row to itself, which would otherwise store
		// one row per language saying what the master already says — and then the
		// master could never be moved again without moving all of them back.
		if state == master {
			continue
		}
		out.ByLanguage[key] = state
	}
	if len(out.ByLanguage) == 0 {
		out.ByLanguage = nil
	}
	if out.Master == "" && out.ByLanguage == nil {
		return "", true
	}
	b, err := json.Marshal(out)
	if err != nil {
		return "", false
	}
	return string(b), true
}

// ---- the scope rung: a work's or a board's own opinion (0073) ---------------

// normalizeTextOrderScope is the whole of what a book, a film, a show, a game or
// a board may store in its `text_order` column — ONE function for all five,
// because "similar things act similarly" and five copies of a four-value
// whitelist is four chances for one of them to drift.
//
// IT IS NOT normalizeTextOrder ABOVE, and the difference is the shape rather than
// the vocabulary. That one round-trips the READER's blob — a master plus a table
// keyed by language, with rows that agree with the master dropped. This one holds
// a single state, because a work has one answer and not a table: a book is not
// bilingual by container the way a reader is by habit.
//
// ” IS INHERIT AND IS ALWAYS ACCEPTED. A work with no opinion stores nothing and
// the ladder falls through to the language and then to the master, which is what
// clearing the control has to mean. Whitespace is trimmed into that same answer,
// so a client sending " " is clearing it rather than storing a state no reader can
// see or remove.
//
// AN UNKNOWN STATE IS REFUSED rather than dropped, which is the rule the blob
// above already follows for the same reason: a value this server has never heard
// of is a client sending something it invented, and storing it would put a card
// into a state the reader cannot reach the control for.
func normalizeTextOrderScope(raw string) (string, bool) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", true
	}
	if !validTextOrder(raw) {
		return "", false
	}
	return raw, true
}

// textOrderList is the error message's half of the same table, so a 400 names the
// four states rather than saying a value was wrong without saying what is right.
func textOrderList() string {
	return strings.Join(textOrders, " / ")
}
